test/Sema/diag_get_vs_set_subscripts.swift

```diff
@@ -0,0 +1,26 @@
+// RUN: %target-typecheck-verify-swift
+
+enum Key: Int {
+  case aKey
+  case anotherKey
+}
+ 
+class sr6175 {
+  var dict: [Key: String] = [:]
+  func what() -> Void {
+    dict[.notAKey] = "something" // expected-error {{cannot subscript a value of type '[Key : String]' with an index of type '_'}}
```

Comment:
This is definitely better than `cannot assign through subscript: subscript is get-only` in this case, but it would be great if instead of `type "_"` we'd actually went and tried to diagnose `UnresolvedMemberExpr` here and diagnose that `.notAKey` is invalid, because it doesn't really help users that much when it's "_"...

Reply:
I agree! That problem is wider that just subscripts, though, it happens for all unresolved params of overloads. E.g.:
```
func f(_ i: Int) -> Void {}
func f(_ d: Double) -> Void {}
f(.notAThing)
```
Which results in a terrible "expression type 'Void' is ambiguous without more context", which is really FailureDiagnosis just giving up entirely. I'd be happy to work on improving this, but it ought to be done for all cases that use CallerCandidateInfo, so maybe a separate PR?

Reply:
Yes indeed, but I think it might be easy win to improve it for `dict` subscript, but maybe not... Current diagnostics are just not suited for certain kind of diagnostics, like this one requiring contextual type, I have couple of ideas how to fundamentally change it by getting more info from constraint solver too, just need to get to this.

Reply:
Ok. It looks to me like the right thing to do ought to be to determine the contextual type for the index expression (Key) and then typeCheckChildIndependently() on that index expression passing in Key as the contextual type, purpose CTP_CallArgument.

And this already happens and works if the chosen subscript index type is something like Int. You get a nice error from the unresolved member expression like "type 'Int' has no member 'notAKey'".

The issue in the original example is that typeCheckChildIndependently() calls validateContextualType() during its setup, which specifically throws away archetype types and so you get nothing out of rechecking. So maybe there needs to be a new typeCheck variant method or flag passed to the existing one which is "no, seriously, this is the contextual type for this expression, don't mess with it"?